### ISSUE TITLE
Metadata for many-to-many relations

### ISSUE THREAD

**Issue by quetzyg** [tobscure/json-api#48]
Hi there!

When dealing with **many-to-many** relations, sometimes there's the need to include data which doesn't belong to any of the **types**, but does belong to the relation they form (the pivot table).

Currently, there's no way to add that kind of data, since the only thing that goes into `relationships` is the `id` and the `type` of any related **type**, as we can see [here](https://github.com/tobscure/json-api/blob/master/src/Elements/Resource.php#L200).

I [asked](http://discuss.jsonapi.org/t/pivot-data-for-many-to-many-relations/113) how we could handle that case in the JSON API discuss, in which I was pointed to this [other question](http://discuss.jsonapi.org/t/metadata-about-relationships/86), from where I got my [answer](http://discuss.jsonapi.org/t/metadata-about-relationships/86/2).

What's the possibility to add that feature to the library?

Cheers,
Quetzy

**Comment by tobscure** [comment]
Thanks for reporting. I've just committed a fix in the 0.2 branch; meta information will be added to resource identifier objects if present. When composing your relationship data, you can either create the Resources beforehand and set the metadata:

```php
public function tags()
{
    return new ClosureHasManyBuilder($serializer = new TagSerializer, function ($post) use ($serializer) {
        return array_map(function ($tag) use ($serializer) {
            $resource = new Resource($tag, $serializer);
            $resource->setMeta($tag->metadata);
            return $resource;
        }, $post->tags);
    });
}
```

... or you can return the raw data and configure the resources afterwards:

```php
public function tags()
{
    $builder = new ClosureHasManyBuilder(new TagSerializer, function ($post) {
        return $post->tags;
    });

    $builder->configure(function (Relationship $relationship) {
        foreach ($relationship->getData()->getResources() as $resource) {
            $resource->setMeta($resource->getData()->metadata);
        }
    });

    return $builder;
}
```

(Note: haven't actually tested this!)

**Comment by quetzyg** [comment]
Brilliant! :+1:
